src/uscheduleedit.1.in
.TH uscheduleedit 1 \*(VE \*(PA      
.hy 0
.URL "Uwe Ohse" ../uwe.html
.br
.URL "The uschedule package" ../uschedule.html
.paragraph
.SH NAME
uscheduleedit - edit a uschedule command
.LINE
.section SYNOPSIS
\fBuscheduleedit\fR [-.] [-d DIR] ID
.section DESCRIPTION
\fBuscheduleedit\fR allows you to edit the registered uschedule command with
the identifier \fIID\fR, using $VISUAL, $EDITOR or vi.
.paragraph
Note that a command may be described by two different files, so don't
be surprised if your editor offers you to edit two files.
.section OPTIONS
.options
.option "-., --dot-as-home"
The current working directory will be used instead of $HOME.
.option "-d, --dir=DIR"
Read jobs from \fIDIR\fR. The default is \fI~/.uschedule/\fR.
.section CAVEAT
If the editor of your choice breaks hard links then
\fBuschedulerm\fR will lose the ability to recognize which registered
commands are still in use. It will then not inform you a command you
want to delete about that and will not stop you from deleting it. The
daemon is able to deal with this, though.
.section AUTHOR
Uwe Ohse, uwe@ohse.de
.section "SEE ALSO"
.manlink uschedule.html uschedule(1) ,
.manlink uschedulecmd.html uschedulecmd(1) ,
.manlink intro.html uschedule_intro(7) .
.HTML <!--
.paragraph
The homepage may be more up-to-date, see
.br
http://www.ohse.de/uwe/uschedule.html.
.HTML -->


src/NEWS
Version 0.7.1, August 2004:

* fixed uschedulelist --short (-s), which used to display the "next" 
  instead of the "last" time.
* "uschedule --every X" (-e) repeats a job with a given interval X.
* The new uscheduleruntimelimit tool allows to limit the run time of a job.
* uscheduleconf makes base directory sticky for daemontools <0.75.
* uschedulelist --show-command (-S) shows the command in addition to the 
  usual list output.
* uschedule accepts + as synonym for / in TIMESPEC.

Version 0.7.0, March 2003:

* implemented -. option ("use . instead of $HOME").
* bugfix, "-" as username or logusername was mistreated.

  *** Impact: the user may have got root rights. ***

  Note: older versions may have left a multilog running as root instead
  of the user. Impact: the user may have got root rights.
  Fix: Go though /service/*uschedule*/log/run and add 
    setuidgid "username" \
  after the line containing the softlimit call and before the line 
  containing run. Afterwards the file should look like this:
    #! /bin/sh
    exec 2>&1
    cd '...SOMEWHERE...' || exit 1
    exec \
    softlimit -m 8000000 -o 400 -p 40 \
    setuidgid USERNAME \
    ./run

  The code containing the bug has been rewritten and simplified.

* uscheduleconf now has a --no-user-change option.

Version 0.6.1, October 2002:

* installed manual pages again.

Version 0.6.0, October 2002:

* uschedulecmd used to steal and misinterpret options given to the child 
  command. Now all programs stop parsing looking for options after the
  first non-option argument.
* implemented start/end times for jobs. See the uschedule manual page
  for more information.
* uschedulelist now prints all month in the 1..12 range, not 0..11 anymore.

Version 0.5.8, November 2001:

* internal changes to get rid of some warnings.
* the same format is used whereever a time is printed.

Version 0.5.7, November 2001:

* uscheduleconf created completely broken configuration if a log account
  different from the user account was given.

Version 0.5.6, October 2001:

* uscheduleconf created buggy call to multilog.

Version 0.5.5, October 2001:

* fixed another self check problem.

Version 0.5.4, October 2001:

* fixed a number of documentation bugs.
* uscheduleconf now starts uscheduled instead of the nonexistant scheduled.
* self check now works if the machine has a time zone different from UTC
  or the local file system has a block size different from 8k.

Version 0.5.3, September 2001:

* first public release.

Version 0.5.1 to 0.5.2, September 2001:

* internal release.

Version 0.5.0, August 2001:

* first internal release.


src/uscheduleedit.c
#include "uogetopt.h"
#include "scheduled.h"
#include "open.h"
#include "uogetopt.h"
#include "stralloc.h"
#include "wait.h"
#include "bailout.h"
#include "pathexec.h"
#include "env.h"
#include "error.h"
#include "uolock.h"
#include "fmt.h"
#include "wrap_stat.h"
#include <unistd.h>

static const char *o_dir;
static int o_dot_as_home;

static uogetopt2 myopts[]={
{'.',"dot-as-home",uogo_flag,UOGO_NOARG,&o_dot_as_home, 1 ,
"Use current directory instead of $HOME.",0,0},
{'d',"dir",uogo_string,0,&o_dir, 0 ,
"Jobdirectory.",
"Jobs will be read from this directory.","DIR"},
{0,0,0,0,0,0,0,0,0}
};              

static int doit(char *vi, char *file1, char *file2)
{
	int pid;
	int code;
	int ret;
	char nb[FMT_ULONG];
	pid=fork();
	if (-1==pid)
		xbailout(111,errno,"failed to fork",0,0,0);
	if (0==pid) {
		char *av[4];
		av[0]=vi;
		av[1]=file1;
		av[2]=file2;
		av[3]=0;
		pathexec(av);
		xbailout(111,errno,"failed to execute ",av[0],0,0);
	}
	ret=wait_pid(&code,pid);
	if (-1==ret) xbailout(111,errno,"failed to wait for ",vi," process",0);
	if (0==ret) xbailout(111,0,"wait_pid returned 0",0,0,0); /* ECANTHAPPEN */
	if (0==code)
		return 0;
	nb[fmt_ulong(nb,code)]=0;
	warning(0,vi, " exited with code ",nb,0);
	return (code & 127) ? 1: code >> 8;
}

static uogetopt_env myoptenv={
"uscheduleedit",PACKAGE,VERSION,
"uscheduleedit [options] ID",
"This program allows you to edit the schedulejob ID.",
"long",
"Report bugs to uschedule@lists.ohse.de",
2,2,0,0,uogetopt_out,myopts
};

int main(int argc, char **argv) 
{
	static stralloc fn;
	static stralloc fn2;
	struct wrap_stat st;
	int fd;
	char *e;

	bailout_progname(argv[0]);
	flag_bailout_fatal_begin=3;
	uogo_posixmode=1;
	myoptenv.program=flag_bailout_log_name;
	uogetopt_parse(&myoptenv,&argc,argv);

	change_dir(0,o_dir,o_dot_as_home);
	check_id(argv[1]);
	if (!stralloc_copys(&fn,IDDIR)) oom();
	if (!stralloc_cats(&fn,"/")) oom();
	if (!stralloc_cats(&fn,argv[1])) oom();
	if (!stralloc_copy(&fn2,&fn)) oom();
	if (!stralloc_0(&fn)) oom();
	if (!stralloc_cats(&fn2,".run")) oom();
	if (!stralloc_0(&fn2)) oom();

	fd=open_write(fn.s);
	if (-1==fd) 
		xbailout(111,errno,"failed to open_write ",fn.s,0,0);
	if (-1==wrap_fstat(fd,&st)) 
		xbailout(111,errno,"failed to fstat ",fn.s,0,0);
	if (-1==uolock_tryexcl(fd))
		xbailout(111,errno,"failed to lock ",fn.s,0,0);

	e=env_get("VISUAL");
	if (!e)
		e=env_get("EDITOR");
	if (!e) {
		static char vi[]="vi";
		e=vi;
	}
	if (-1==wrap_stat(fn2.s,&st)) {
		if (errno!=error_noent)
			xbailout(111,errno,"failed to stat ",fn2.s,0,0);
		doit(e,fn.s,0);
	} else
		doit(e,fn.s,fn2.s);
	return 0;
}
